### PR TITLE
Rewrite docker_container to use Docker API directly

### PR DESCRIPTION
**Pull request by felixfontein**
##### SUMMARY
My last planned big part of #364: rewrite docker_container to use the Docker API directly (instead of Docker SDK for Python). It's not done yet, but it already passes all existing integration tests (at least it does so locally, let's see what CI says) :-) Some parts of it are very WIP, some already in a good state IMO.

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
docker_container

### CODE REVIEW

**Comment by felixfontein** [comment]
ready_for_review

**Comment by felixfontein** [comment]
If you are interested in how this can be extended (i.e. new options added), see for example https://github.com/ansible-collections/community.docker/pull/427/commits/0cb6c20512ed03e33381eddb05ce9b6ee6027a83 (part of #427) or https://github.com/ansible-collections/community.docker/pull/426/commits/9577160ea61c51a05e1adc682b396b5523484679 (part of #426).

File: plugins/module_utils/module_container/docker_api.py [hunk 1]
```diff
@@ -0,0 +1,1327 @@
...
+            values['published_ports'][port] = tuple([default_ip] + list(port_spec[1:]))
+    return values
+
+
+def _preprocess_container_names(module, client, api_version, value):
+    if value is None or not value.startswith('container:'):
+        return value
+    container_name = value[len('container:'):]
+    # Try to inspect container to see whether this is an ID or a
+    # name (and in the latter case, retrieve it's ID)
```
Comment (briantist) [inline plugins/module_utils/module_container/docker_api.py:1160]: ultra nit
```suggestion
    # name (and in the latter case, retrieve its ID)
```

**Review by briantist — COMMENTED** [review]
I have only medium confidence that I was able to review this thoroughly; rewrites are a lot haha, but it looks good to me, the tests seem to have good coverage, though some are failing at the moment (seem to be catching something relevant so that's good).

File: changelogs/fragments/docker_container.yml [hunk 2]
```diff
@@ -0,0 +1,11 @@
...
+  - "docker_container - the module was completely rewritten from scratch (https://github.com/ansible-collections/community.docker/pull/...)."
+breaking_changes:
+  - "docker_container - ``publish_all_ports`` is no longer ignored in ``comparisons`` (https://github.com/ansible-collections/community.docker/pull/...)."
+  - "docker_container - ``exposed_ports`` is no longer ignored in ``comparisons``. Before, its value was assumed to be identical with the value of ``published_ports`` (https://github.com/ansible-collections/community.docker/pull/...)."
+  - "docker_container - ``log_options`` can no longer be specified when ``log_driver`` is not specified (https://github.com/ansible-collections/community.docker/pull/...)."
+  - "docker_container - ``restart_retries`` can no longer be specified when ``restart_policy`` is not specified (https://github.com/ansible-collections/community.docker/pull/...)."
+  - "docker_container - ``stop_timeout`` is no longer ignored for idempotency if told to be not ignored in ``comparisons``. So far it defaulted to ``ignore`` there, and setting it to ``strict`` had no effect (https://github.com/ansible-collections/community.docker/pull/...)."
```
Comment (gotmax23) [inline changelogs/fragments/docker_container.yml:11]: Don't forget to replace these with the actual URL to the PR

File: plugins/module_utils/module_container/base.py [hunk 3]
```diff
@@ -0,0 +1,1185 @@
+# Copyright (c) 2022 Felix Fontein <felix@fontein.de>
+# Copyright 2016 Red Hat | Ansible
```
Comment (gotmax23) [inline plugins/module_utils/module_container/base.py:2]: Why would there be a copyright from 2016 if this was completely rewritten? Or was it not?

Comment (felixfontein) [reply plugins/module_utils/module_container/base.py:2]: This contains some code from the old version, so the old copyright needs to stay there AFAIK.

File: changelogs/fragments/docker_container.yml (hunk 2 above)
Comment (felixfontein) [reply changelogs/fragments/docker_container.yml:11]: Oooops, thanks for noticing!

Comment (felixfontein) [reply changelogs/fragments/docker_container.yml]: Fixed in 89bf3da6c763c7d95378d54ad3c681fba3992cec.

File: plugins/module_utils/module_container/base.py (hunk 3 above)
Comment (gotmax23) [reply plugins/module_utils/module_container/base.py:2]: > This contains some code from the old version, so the old copyright needs to stay there AFAIK.

That's correct. I'll resolve this.

**Comment by felixfontein** [comment]
@briantist @gotmax23 thanks a lot for reviewing this!

**Comment by gotmax23** [comment]
> @briantist @gotmax23 thanks a lot for reviewing this!

Sure! I'm not very familiar with writing ansible plugins (yet!), but happy I was able to help.

**Pull request merged**